docs/4.4/api-reference.md

```diff
@@ -5,63 +5,827 @@ description: The detailed guide to Teleport API
+
+- You want to write a program that can always ensure certain roles exist on your system and do not want to orchestrate `tctl` to do this.
+- You want to dynamically create short lived roles.
+- You want to dynamically create roles with fields filled that Teleport currently does not support.
+
+### The Role Object
+
+A Teleport `role` is defined by its `Allow` rules, `Deny` rules, and OpenSSH `Options`. We'll break these down piece by piece below.
+
+To see a role example in `yaml` form, look at the the admin role in the [RBAC documentation](enterprise/ssh-rbac.md). You'll notice that the role object below has the same exact nested structure as its `yaml` counterpart.
+
+```go
+// RoleV3 represents role resource specification
+type RoleV3 struct {
+  // Resource Fields are fields that all Teleport resources have - see above
+  Resource Fields
+  // Spec is the role specification
+  Spec RoleSpecV3 struct {
+    // Options is for OpenSSH options like agent forwarding.
+    Options RoleOptions
+    // Allow is the set of conditions evaluated to grant access.
+    Allow RoleConditions
+    // Deny is the set of conditions evaluated to deny access. Deny takes priority over allow.
+    Deny RoleConditions
+  }
+}
+```
+
+**Role Options**
+
+The `RoleOptions` struct defines what OpenSSH actions a user is allowed to use.
+
+```go
+// RoleOptions is a set of role options
+type RoleOptions struct {
+  // ForwardAgent is SSH agent forwarding. default true.
+  ForwardAgent Bool
+  // MaxSessionTTL defines how long a SSH session can last for.
+  MaxSessionTTL Duration
+  // PortForwarding defines if the certificate will have "permit-port-forwarding"
+  // in the certificate. PortForwarding is true if not set.
+  PortForwarding *BoolOption // struct { Value bool } - Nullable boolean
+  CertificateFormat string
+  // ClientIdleTimeout sets disconnect clients on idle timeout behavior,
+  // if set to 0 means do not disconnect, otherwise is set to the idle duration.
+  ClientIdleTimeout Duration
+  // DisconnectExpiredCert sets disconnect clients on expired certificates.
+  DisconnectExpiredCert Bool
+  // BPF defines what events to record for the BPF-based session recorder.
+  BPF []string
+  // PermitX11Forwarding authorizes use of X11 forwarding.
+  PermitX11Forwarding Bool
+  // MaxConnections defines the maximum number of
+  // concurrent connections a user may hold.
+  MaxConnections int64
+  // MaxSessions defines the maximum number of
+  // concurrent sessions per connection.
+  MaxSessions int64
+}
+```
+
+**Role Conditions**
+
+The `RoleConditions` struct allows for precise permission combinations between a role's `Allow` and `Deny` fields.
+
+`Deny` conditions take priority over `Allow` conditions. If a user has multiple roles, their role permissions will be evaluated together, which can potentially lead to conflicting conditions. This can result in a user being denied permissions that another role of theirs allows.
```

Comment:
I would describe it something like below:

Deny conditions are evaluated first and they are logically OR'ed. That means if any role matches a deny section for either labels or logins, you will be denied access. If you are not denied access to anything, then the allow conditions are evaluated and they are logically AND'ed. That means you have to match both labels and logins to be allowed access.

This is the function that does the evaluation.

https://github.com/gravitational/teleport/blob/master/lib/services/role.go#L1719-L1774